static/js/components/dashboard/courses/StatusMessages.js

```diff
@@ -0,0 +1,227 @@
+import { COURSE_CARD_FORMAT } from '../../../constants';
+
+type Message = {
+  message: string|React$Element<*>,
+  action?: React$Element<*>
+};
+
+const makeCouponMessage = coupon => (
+  `${couponMessageText(coupon)}${makeCouponReason(coupon)}.`
+);
+
+export const formatAction = (message: Message) => (
+  <div className="second-col">
+    { message.action }
+  </div>
+);
+
+export const formatMessage = (message: Message, index?: number) => (
+  <div className="status-message cols" key={index}>
+    <div className="message first-col">
+      { message.message }
+    </div>
+    { message.action ? formatAction(message) : null }
+  </div>
+);
+
+const formatDate = date => (
+  moment(date).format(COURSE_CARD_FORMAT)
+);
+
+type CalculateMessagesProps = {
+  courseAction:                (run: CourseRun, actionType: string) => React$Element<*>,
+  firstRun:                    CourseRun,
+  course:                      Course,
+  expandedStatuses:            Set<number>,
+  setShowExpandedCourseStatus: (n: number) => void,
+  coupon?:                     Coupon,
+};
+
+// this calculates any status messages we'll need to show the user
+// we wrap the array of messages in a Maybe, so that we can indicate
+// the case where there are no messages cleanly
+// 
+// I'm so sorry for anyone who has to read / modify this
+// sometimes there really isn't a better way :/
+export const calculateMessages = (props: CalculateMessagesProps) => {
+  const {
+    coupon,
+    courseAction,
+    firstRun,
+    course,
+    expandedStatuses,
+    setShowExpandedCourseStatus,
+  } = props;
+
+  let exams = course.has_exam;
+  let paid = firstRun.has_paid;
+  let passedExam = hasPassingExamGrade(course);
+  let paymentDueDate = moment(firstRun.course_upgrade_deadline);
+
+  if (firstRun.status === STATUS_PAID_BUT_NOT_ENROLLED) {
+    const contactHref = `mailto:${SETTINGS.support_email}`;
+    return S.Just([{
+      message: <div>
+        {"Something went wrong. You paid for this course but are not enrolled. "}
+        <a href={contactHref}>
+          Contact us for help.
+        </a>
+      </div>
+    }]);
+  }
+
+  // Course run isn't enrollable, user never enrolled
+  if (!isEnrollableRun(firstRun) && !R.any(userIsEnrolled, course.runs)) {
+    return S.Just([{
+      message: "There are no future course runs scheduled."
+    }]);
+  }
+
+  // Course is running, user has already paid, we show no messages
+  if (courseUpcomingOrCurrent(firstRun) && paid) {
+    return S.Nothing; 
+  }
+
+  let messages = [];
+
+
+  if (coupon && coupon.content_type === COUPON_CONTENT_TYPE_COURSE && coupon.object_id === course.id) {
+    messages.push({ message: makeCouponMessage(coupon)});
+  }
+
+  // handle other 'in-progress' cases
+  if (courseUpcomingOrCurrent(firstRun) || firstRun.status === STATUS_CAN_UPGRADE) {
+    if (hasPaidForAnyCourseRun(course)) {
+      messages.push({message: "You are auditing this course."});
```

Comment:
Hm, I'm not sure exactly what the situation is, but I think I remember Rob explaining to me that it's not uncommon for students to re-take courses after they have already paid and passed the course because they're acting as TAs. So my understanding was that if they pass and pay for the course once we could that as a pass regardless of whether they re-take the course (although if they wanted to pay to get a better grade for some reason they would need to). does that make sense? I'm always getting this kind of stuff third hand so I really don't know, haha.